### PR TITLE
[fpmsyncd] Fpmsyncd Next Hop Table Enhancement

### PR DESCRIPTION
**Pull request by ntt-omw**
#### Why I did it
Implementing code changes for https://github.com/sonic-net/SONiC/pull/1425

#### How I did it
Added RTA_NH_ID attribute patch to libnl3 to add nexthop group feature to fpmsyncd.

##### enable/disable nexthop group feature

- Klish will call REST API to configure `feature next-hop-group enable`.
- `FEATURE|nexthop_group` will be created in `CONFIG_DB`
  - template `zebra.conf.j2` will generate `zebra.conf` with `fpm use-next-hop-groups` if `FEATURE|nexthop_group` exists in `CONFIG_DB`. Else, it will generate `zebra.conf` with `no fpm use-next-hop-groups` (default behavior)
- Do `config save` comman and write to `/etc/sonic/config_db.json`
- restart SONiC: `virsh reboot sonic-nhg`
- `/etc/frr/zebra.conf` has `fpm use-next-hop-groups` instead of `no fpm use-next-hop-groups`

##### Klish CLI for feature nexthop_group

- Enable: `sonic(config)# feature next-hop-group enable`
- Disable: `sonic(config)# no feature next-hop-group`

Enable

```sh
admin@sonic:~$ sonic-cli

sonic# configure terminal

sonic(config)#
  end        Exit to EXEC mode
  exit       Exit from current mode
  feature    Configure additional feature
  interface  Select an interface
  ip         Global IP configuration subcommands
  mclag      domain
  no         To delete / disable commands in config mode

sonic(config)# feature
  next-hop-group  Next-hop Groups feature

sonic(config)# feature next-hop-group
  enable  Enable Next-hop Groups feature

sonic(config)# feature next-hop-group enable
  <cr>
```

Disable

```sh
sonic(config)# no
  feature  Disable additional feature
  ip       Global IP configuration subcommands
  mclag    domain

sonic(config)# no feature
  next-hop-group  Disable Next-hop Groups feature

sonic(config)# no feature next-hop-group
  <cr>
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2 [hunk 1]
```diff
@@ -7,8 +7,13 @@
 {% endblock banner %}
 !
 {% block fpm %}
+{% if ( ('nexthop_group' in FEATURE) and ('state' in  FEATURE['nexthop_group']) and
+        (FEATURE['nexthop_group']['state'] == 'enabled') ) %}
+fpm use-next-hop-groups
```
Comment (lguohan) [inline dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2:12]: can you follow example here?

https://github.com/sonic-net/SONiC/blob/master/doc/BGP/BGP-supress-fib-pending.md#632--device_metadata

Comment (ebiken) [reply dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2:12]: @lguohan Moved under device metadata following the example.

**Comment by nakano-omw** [comment]
please help in restarting the /AzurePipelines run Azure.sonic-buildimage.
Would you be able to grant me the privilege?

**Comment by nakano-omw** [comment]
@lguohan 
Please help in restarting the /AzurePipelines run Azure.sonic-buildimage. Would you be able to grant me the privilege?

**Comment by ridahanif96** [comment]
@nakano-omw you need to repush your code.

**Comment by nakano-omw** [comment]
@ridahanif96 Thanks for the comment. Fix conflicts and build test before repush.

**Comment by nakano-omw** [comment]
@ridahanif96 @zhangyanzhao It has been reviewed in Routing WG. Could we approve and merge the PR ?

**Comment by zhangyanzhao** [comment]
@lguohan  can you please merge this PR if you have no more comments? Thanks.

**Comment by nakano-omw** [comment]
@lguohan can you please merge this PR. Thanks.

**Comment by dgsudharsan** [comment]
@ntt-omw  Can you please sync to the latest. We re enabled nexthop kernel support https://github.com/sonic-net/sonic-buildimage/pull/18953

File: src/sonic-yang-models/yang-models/sonic-device_metadata.yang [hunk 2]
```diff
@@ -254,6 +254,16 @@ module sonic-device_metadata {
                     type inet:ipv4-address;
                     description "BGP Router identifier";
                 }
+
+                leaf nexthop_group {
```
[the remark this answers is not in the document]
Comment (nakano-omw) [reply src/sonic-yang-models/yang-models/sonic-device_metadata.yang:258]: I added the sample_config_db. Thanks.

**Review by dgsudharsan — APPROVED** [review]

**Review by eddieruan-alibaba — APPROVED** [review]

**Comment by eddieruan-alibaba** [comment]
Cherry picked into phoenixwing folk.http://phoenixwing.com.cn/codes

**Comment by kperumalbfn** [comment]
@ntt-omw please update the branch and resolve the conflicts.

**Comment by nakano-omw** [comment]
@kperumalbfn @eddieruan-alibaba @qiluo-msft  Branch updated and conflicts resolved. Please approve and merge.

**Comment by eddieruan-alibaba** [comment]
Thanks @nakano-omw  I have reviewed it again. My approval is still there.

**Comment by nakano-omw** [comment]
@qiluo-msft It is marked as needing review by qiluo-msft. This PR has been reviewed by the Routing WG. Please approve.

File: dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2 [hunk 3]
```diff
@@ -7,8 +7,14 @@
 {% endblock banner %}
 !
 {% block fpm %}
+{% if ( ('localhost' in DEVICE_METADATA) and ('nexthop_group' in  DEVICE_METADATA['localhost']) and
+        (DEVICE_METADATA['localhost']['nexthop_group'] == 'enabled') ) %}
+! enable next hop group support
+fpm use-next-hop-groups
+{% else %}
```
[the remark this answers is not in the document]
Comment (kentaro-arrcus) [reply dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2:14]: Thanks for the review. But I'm away from this project now. Hi @nakano-omw , could you please follow up with @kperumalbfn ? Thanks.

[the remark this answers is not in the document]
Comment (nakano-omw) [reply dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2:14]: @kperumalbfn Thanks for the review. The sonic-mgmt tests are being created in the Phoenixwing .
https://github.com/sonic-net/sonic-mgmt/pull/13785

[the remark this answers is not in the document]
Comment (kperumalbfn) [reply dockers/docker-fpm-frr/frr/zebra/zebra.conf.j2:14]: Thanks @nakano-omw Will review sonic-mgmt tests as well.

**Pull request merged**